hd-wallet/pluginv2.lua

```diff
@@ -0,0 +1,602 @@
+--   "xpub": "xpub68Z371jLPJn11bzL9rYWhKgE6ysTPx2DpMkybSnmGUUYvgjT68yvYoCw6PP8Vo7YoZRC6iqrfpixEUG694KgHPYYnydGuEYDwjESStYxYxe",
+--   "signature": "3045022100af9bf94c4959328b56861ca5f175b5e59014cb5bd2a5fcee2e95b1563dbc652e0220411ff01751af64d6b7209908fc58f527b07a0a9258eee7be7aa5704136954b02",
+--   "coin_signature": "af9bf94c4959328b56861ca5f175b5e59014cb5bd2a5fcee2e95b1563dbc652e411ff01751af64d6b7209908fc58f527b07a0a9258eee7be7aa5704136954b02"
+-- }
+-- ```
+-- 
+-- ## Example Input JSON object for ETH
+-- 
+-- ```
+-- {
+--    "masterKeyId": "5aef3e3f-7927-49b2-b252-bf84b6980f95",
+--    "coin": "eth",
+--    "msgHash": "45a0ee821b05400f513891bbb567a99139f3df72e9e1d4b48186841cc5996d2f"
+-- }
+-- ```
+--
+-- ## Example Input JSON object for XPRV 
+-- ```
+-- {
+--    "import": true,
+--    "xprvEncId": "5aef3e3f-7927-49b2-b252-bf84b6980f95",
+--    "wrapKeyId": "02e5c697-87cd-45b6-8163-957fa5e13370",
+--    "name": "Imported Key"
+-- }
+-- ```
+--
+-- ## References
+-- 
+-- - https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki
+-- - https://xrpl.org/sign.html
+-- 
+-- ### Release Notes
+--  - Initial release
+
+---------- @@@@@@@@@@@@@@@@@@@@@@@@@ ----------
+----------      COMMON UTILITIES     ----------
+---------- @@@@@@@@@@@@@@@@@@@@@@@@@ ----------
+----------------- Debug Logging ---------------
+local DEBUG = {}
+local DEBUG_N = 1
+function debug(msg)
+  DEBUG[DEBUG_N] = msg
+  DEBUG_N = DEBUG_N + 1
+end
+
+----------------- Constant --------------------
+local PRIVATE_WALLET_VERSION =  "0488ADE4"
+local PUBLIC_WALLET_VERSION = "0488B21E"
+local FIRST_HARDENED_CHILD = 2147483648
+local PUBLIC_KEY_COMPRESSED_LENGTH = 33
+local N = "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141"
+
+------------- BIP32 key structure -------------
+local key = {["version"]="",
+    ["depth"]="",       -- 1 byte
+    ["index"]="",       -- 4 byte child number
+    ["fingerprint"]="", -- 4 byte parent fingerprint
+    ["chain_code"]="",  -- 32 byte
+    ["key"]="",         -- 33 byte long key
+    ["checksum"]="",    -- checksum of all above
+    ["is_private"] = ""} -- 1 bit flag
+
+function create_ASN1_private_key(key_byte)
+  while string.len(key_byte) < 64 do
+    key_byte = '00' .. key_byte
+  end
+  return "302E0201010420".. key_byte .."A00706052B8104000A"
+end
+
+-- return hex of exported key
+function decode_key(exported_master_key_serialized)
+  local blob = Blob.from_base58(exported_master_key_serialized)
+  return blob:hex()
+end
+
+-- deserialize bip32 key
+function deserialize(exported_master_key_serialized)
+    hex_key = decode_key(exported_master_key_serialized)
+
+    key.version = string.sub(hex_key, 1, 8)
+    key.depth = string.sub(hex_key, 9, 10)
+    key.index = string.sub(hex_key, 11, 18)
+    key.fingerprint = string.sub(hex_key, 19, 26)
+    key.chain_code = string.sub(hex_key, 27, 90)
+    key.key = string.sub(hex_key, 91, 156)
```

Comment:
I think this is a good opportunity to use Blobs everywhere instead of strings, specially for secret bytes.
Same for all functions handling private bytes, e.g. `decode_key` could simply return the blob, and `create_ASN1_private_key` concatenates, etc.

Reply:
Hi @zugzwang thanks for the comments.  I'd like to avoid changes here if possible as this is production code and we don't want to upstream any defects as part of this exercise.  Could you help me understand the utility of moving from strings to bytes across this plugin?

Reply:
Hello. 
Very generally, it is better to avoid touching secret bytes as most as possible, and that includes conversions (which, for instance, could be non-constant time). The lifetime of secret bytes (keys, messages) should only involve creation from randomness and cryptographic usage. So in this plugin, I noticed that the conversion to strings is unnecessary. 
Also, there could be efficiency improvements if the plugin is used in high volume.

Reply:
@zugzwang our `Blob` type is a wrapper around `string` in Lua. If we had a special type for cryptographic secrets that allowed constant-time manipulations then we could use that here, but unfortunately we don't have such type in our Lua ecosystem.

Reply:
Hello @mzohreva, thanks for the clarification!

@jordan-rego-erisx, It still looks to me that these conversions are unnecessary and the code would be simpler without. Since I'm the one bugging about this, I can do so in a subsequent PR.

I'll review the rest of the PR today.

Reply:
@zugzwang thanks!  Happy to discuss at the meeting later too